src/lib.rs

```diff
@@ -71,6 +72,104 @@ impl fmt::Display for Error {
+// instructions.
+pub const PICTURE_ALIGNMENT: usize = 64;
+
+/// Allocation for a picture.
+#[derive(Debug, PartialEq, Eq)]
+pub struct PictureAllocation<D: Send + 'static> {
+    /// Pointers to planar image data.
+    ///
+    /// Y is `data[0]`, U is `data[1]`, V is `data[2]`.
+    ///
+    /// The data should be `u8` (for 8 bits) or `u16` (for 10 bits).
+    ///
+    /// The `data[0]`, `data[1]` and `data[2]` must be [`PICTURE_ALIGNMENT`] byte aligned and with a pixel
+    /// width/height multiple of 128 pixels. Any allocated memory area should also be padded by
+    /// [`PICTURE_ALIGNMENT`] bytes.
+    ///
+    /// `data[1]` and `data[2]` must share the same `stride[1]`.
+    pub data: [*mut u8; 3],
+
+    /// Number of bytes between 2 lines in `data[]` for luma in case of `stride[0]` or chroma in
+    /// case of `stride[1]`.
+    pub stride: [isize; 2],
+
+    /// Allocator data that can be retrieved back from the picture later.
+    pub allocator_data: D,
+}
+
+unsafe impl<D: Send + 'static> Send for PictureAllocation<D> {}
+
+pub trait PictureAllocator: Send + Sync + 'static {
+    /// Allocator data that is stored together with the [`Picture`].
+    ///
+    /// This can be retrieved from the picture via [`Picture::allocator_data()`].
+    type AllocatorData: Send + 'static;
+
+    /// Allocate the picture buffer based on `pic_params`.
+    ///
+    /// See the [`PictureAllocation`] documentation for the requirements on the allocated memory.
+    ///
+    /// This function will be called on the main thread (the thread which calls
+    /// [`Decoder::get_picture()`]).
+    ///
+    /// # Safety
+    ///
+    /// This function needs to allocate enough memory with the constraints outlined in the
+    /// [`PictureAllocation`] documentation.
+    unsafe fn alloc_picture(
+        &self,
+        pic_params: &PictureParameters,
+    ) -> Result<PictureAllocation<Self::AllocatorData>, Error>;
+
+    /// Release the picture buffer.
+    ///
+    /// If frame threading is used, this function may be called by the main
+    /// thread (the thread which calls [`Decoder::get_picture()`]) or any of the frame
+    /// threads and thus must be thread-safe. If frame threading is not used,
+    /// this function will only be called on the main thread.
+    ///
+    /// # Safety
+    ///
+    /// This function needs to release the memory in `allocation` and can assume that it
+    /// corresponds to a previous call to [`PictureAllocator::alloc_picture()`].
+    unsafe fn release_picture(&self, allocation: PictureAllocation<Self::AllocatorData>);
+}
+
+/// Default allocator.
+///
+/// Note that this allocator is not functional. Calling any of its methods will panic.
+impl PictureAllocator for () {
+    type AllocatorData = ();
+
+    unsafe fn alloc_picture(
+        &self,
+        _pic_params: &PictureParameters,
+    ) -> Result<PictureAllocation<Self::AllocatorData>, Error> {
+        unimplemented!()
```

Comment:
I guess once there is one implementation and tests it will be good to land.

Reply:
That's actually not so easy. We could provide one by implement it all ourselves, but the actual default allocator is not exposed by the C library in a useful way that can be called from the outside. The actual default allocator is also using a memory pool, which would have to be re-implemented here then.

Note that the implementation on `()` is not actually used for anything. We just need something to put into the `<>` for the default case.

Reply:
Done.

> but the actual default allocator is not exposed by the C library in a useful way that can be called from the outside

You might mention `dav1d_default_picture_alloc()` now but this can't be called from the outside. It expects the special initialization in `dav1d_open()` to happen, and that uses private API and can't be replicated from the outside.